aws/ec2.go

```diff
@@ -0,0 +1,301 @@
+			id:               subnetID,
+			availabilityZone: az,
+			public:           isPublic,
+			tags:             convertEc2Tags(sn.Tags),
+		}
+	}
+	return ret, nil
+
+}
+
+func getRouteTables(svc ec2iface.EC2API, vpcID string) ([]*ec2.RouteTable, error) {
+	params := &ec2.DescribeRouteTablesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("vpc-id"),
+				Values: []*string{
+					aws.String(vpcID),
+				},
+			},
+		},
+	}
+	resp, err := svc.DescribeRouteTables(params)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.RouteTables, nil
+}
+
+// Copied from Kubernete's https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go
+func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) (bool, error) {
+	var subnetTable *ec2.RouteTable
+	for _, table := range rt {
+		for _, assoc := range table.Associations {
+			if aws.StringValue(assoc.SubnetId) == subnetID {
+				subnetTable = table
+				break
+			}
+		}
+	}
+
+	if subnetTable == nil {
+		// If there is no explicit association, the subnet will be implicitly
+		// associated with the VPC's main routing table.
+		for _, table := range rt {
+			for _, assoc := range table.Associations {
+				if aws.BoolValue(assoc.Main) == true {
+					subnetTable = table
+					break
+				}
+			}
+		}
+	}
+
+	if subnetTable == nil {
+		return false, fmt.Errorf("could not locate routing table for subnet %q", subnetID)
+	}
+
+	for _, route := range subnetTable.Routes {
+		// There is no direct way in the AWS API to determine if a subnet is public or private.
+		// A public subnet is one which has an internet gateway route
+		// we look for the gatewayId and make sure it has the prefix of igw to differentiate
+		// from the default in-subnet route which is called "local"
+		// or other virtual gateway (starting with vgv)
+		// or vpc peering connections (starting with pcx).
+		if strings.HasPrefix(aws.StringValue(route.GatewayId), "igw") {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func getIntanceTags(ec2Service *ec2.EC2, instanceID string) (map[string]string, error) {
+	params := &ec2.DescribeTagsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("resource-id"),
+				Values: []*string{
+					aws.String(instanceID),
+				},
+			},
+		},
+	}
+	resp, err := ec2Service.DescribeTags(params)
+	if err != nil {
+		return nil, err
+	}
+	tags := make(map[string]string, len(resp.Tags))
+	for _, tagDescription := range resp.Tags {
+		tags[aws.StringValue(tagDescription.Key)] = aws.StringValue(tagDescription.Value)
+	}
+
+	return tags, nil
+}
+
+func convertEc2Tags(instanceTags []*ec2.Tag) map[string]string {
+	tags := make(map[string]string, len(instanceTags))
+	for _, tagDescription := range instanceTags {
+		tags[aws.StringValue(tagDescription.Key)] = aws.StringValue(tagDescription.Value)
+	}
+	return tags
+}
+
+func getClusterName(instanceTags map[string]string) string {
+	if cn, has := instanceTags[kubernetesClusterTag]; has && len(cn) > 0 {
```

Comment:
The comment that follows was the reason for it to be like this. Can be changed afterwards. I'll make it simpler now